docs/src/explanations/dpi.md

```diff
@@ -0,0 +1,143 @@
+---
+layout: docs
+title:  "Calling Native Functions from Chisel (DPI)"
+section: "chisel3"
+---
+
+# Calling Native Functions from Chisel (DPI)
+
+## DPI Basics
+
+DPI API allows you to integrate native code into your Chisel hardware designs. This enables you to leverage existing libraries or implement functionality that is difficult to express directly in Chisel.
+
+Here's a simple example that demonstrates printing a message from a C++ function:
+```c++
+extern "C" void hello()
+{
+    std::cout << "hello from c++\\n";
+}
+```
+
+To call this function from Chisel, we need to define a corresponding DPI object.
+
+```scala
+object Hello extends DPIVoidFunctionImport {
+  override val functionName = "hello"
+  override val clocked = true
+  final def apply() = super.call()
+}
+
+Hello() // Print
+```
+
+Explanation:
+
+* `Hello` inherits from `DPIVoidFunctionImport` because the C++ function doesn't return a value (void).
+* `functionName` specifies the C-linkage name of the C++ function.
+* `clocked = true` indicates that its function call is invoked at clock's posedge.
+* We recommend defining the apply method for a more Scala-like syntax.
+
+## Type ABI
+
+Unlike normal Chisel compilation flow, we use a specific ABI for types to interact with DPI.
+
+### Argument Types
+
+* Operand and result types must be passive.
+* A vector is lowered to an *unpacked* *open* array type, e.g., `a: Vec<4, UInt>` to `byte a []`.
+* A bundle is lowered to a packed struct.
+* Integer types are lowered into 2-state types.
+Small integer types (< 64 bit) must be compatible with C-types and arguments are passed by value. Users are required to use specific integer types for small integers shown in the table below. Large integers are lowered to bit and passed by reference.
+
+
+| Width | Verilog Type | Argument Passing Modes |
+| ----- | ------------ | ---------------------- |
+| 1     | bit          | value                  |
+| 8     | byte         | value                  |
+| 16    | shortint     | value                  |
+| 32    | int          | value                  |
+| 64    | longint      | value                  |
+| > 64  | bit [w-1:0]  | reference              |
+
+### Function Types
+The type of DPI object you need depends on the characteristics of your C++ function:
```

Comment:
The type of DPI object I need for what?

Reply:
Reading ahead, I think this section would be better typed out in prose, rather than summarized as a bulleted list.

"There are several base classes which are used to define DPI functions in Chisel. Which one you use depends on two factors: the return type of the function in C++ and how the DPI is called in relation to the clock."

Then outline the information in the bullet points in two paragraphs after that.